cub/cub/thread/thread_reduce.cuh

```diff
@@ -43,199 +41,515 @@
+_CCCL_NODISCARD _CCCL_DEVICE constexpr bool enable_sm70_simd_reduction()
+{
+  using cub::detail::is_one_of;
+  using ::cuda::std::is_same;
+#  if defined(_CCCL_HAS_NVFP16)
+  return is_same<T, __half>::value && is_one_of<ReductionOp, cub::Sum, cub::Mul>() && Length >= 4;
+#  else
+  return false;
+#  endif
+}
+
+template <typename Input, typename ReductionOp, typename AccumT>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE _CCCL_CONSTEXPR_CXX14 bool enable_simd_reduction()
+{
+  using cub::detail::is_one_of;
+  using ::cuda::std::is_same;
+  using T = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>;
+  if _CCCL_CONSTEXPR_CXX17 (!is_same<T, AccumT>::value)
+  {
+    return false;
+  }
+  else
+  {
+    constexpr auto length = cub::detail::static_size_v<Input>();
+    // clang-format off
+    _NV_TARGET_DISPATCH(
+      NV_PROVIDES_SM_90,
+        (return enable_sm90_simd_reduction<T, ReductionOp, length>() ||
+                enable_sm80_simd_reduction<T, ReductionOp, length>() ||
+                enable_sm70_simd_reduction<T, ReductionOp, length>();),
+      NV_PROVIDES_SM_80,
+        (return enable_sm80_simd_reduction<T, ReductionOp, length>() ||
+                enable_sm70_simd_reduction<T, ReductionOp, length>();),
+      NV_PROVIDES_SM_70,
+        (return enable_sm70_simd_reduction<T, ReductionOp, length>();),
+      NV_IS_DEVICE,
+        (static_cast<void>(length); // maybe unused
+         return false;)
+    );
+    // clang-format on
+    return false;
+  }
+  return false; // nvcc 11.x warning workaround
+}
 
 template <typename Input, typename ReductionOp, typename AccumT>
-_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE // clang-format off
-constexpr bool enable_dpx_reduction()
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE _CCCL_CONSTEXPR_CXX14 bool enable_ternary_reduction()
 {
-  using T = decltype(::cuda::std::declval<Input>()[0]);
-  // TODO: use constexpr variable in C++14+
-  using Lenght = ::cuda::std::integral_constant<int, detail::static_size<Input>()>;
-  return ((Lenght{} >= 9 && detail::are_same<ReductionOp, cub::Sum/*, std::plus<T>*/>()) || Lenght{} >= 10)
-            && detail::are_same<T, AccumT>()
-            && detail::is_one_of<T, int16_t, uint16_t>()
-            && detail::is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum/*, std::plus<T>*/>();
+  using cub::detail::is_one_of;
+  using ::cuda::std::is_same;
+  using T               = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>;
+  constexpr auto length = cub::detail::static_size_v<Input>();
+  if _CCCL_CONSTEXPR_CXX17 (length < 6)
+  {
+    return false;
+  }
+  else
+  {
+    // clang-format off
+    NV_DISPATCH_TARGET(
+      NV_PROVIDES_SM_90,
+        (return (is_one_of<T, ::cuda::std::int32_t, ::cuda::std::uint32_t, ::cuda::std::int64_t, ::cuda::std::uint64_t>
+                 && is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum, cub::BitAnd, cub::BitOr, cub::BitXor>())
```

Comment:
We should move that into a proper variable tempalte or struct

Reply:
uhmm, not sure what you are referring to. It looks pretty readable to me